dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/SurvivalAnalyzer.java

```diff
@@ -0,0 +1,201 @@
+
+      val intervals = compute(filteredDonors);
+      analysis.getResults().add(analysis.new Result(id, intervals));
+    }
+
+    return analysis;
+  }
+
+  private static List<Interval> compute(SearchHit[] donors) {
+    int[] time = new int[donors.length];
+    boolean[] censured = new boolean[donors.length];
+
+    for (int i = 0; i < donors.length; i++) {
+      val donor = donors[i];
+      time[i] = (int) donor.field("donor_survival_time").getValue();
+      censured[i] = ((String) donor.field("donor_vital_status").getValue()).equalsIgnoreCase("alive");
+    }
+
+    val intervals = new ArrayList<Interval>();
+    int startTime = 0;
+    int endTime = 0;
+
+    for (int i = 0; i < time.length; i++) {
+      endTime = time[i];
+      if (censured[i] == false && endTime > startTime) {
+        intervals.add(new Interval(startTime, endTime));
+        startTime = endTime;
+      }
+    }
+    if (endTime > startTime) {
+      intervals.add(new Interval(startTime, endTime));
+    }
+
+    // init variables. Initially everyone is at risk, and the cumulative survival is 1
+    float atRisk = time.length;
+    float cumulativeSurvival = 1;
+    val intervalIter = intervals.iterator();
+
+    // This implementation later mutates this reference.
+    Interval currentInterval = intervalIter.next();
+    currentInterval.setCumulativeSurvival(cumulativeSurvival);
+
+    for (int i = 0; i < time.length; i++) {
+
+      long t = time[i];
+
+      // If we have moved past the current interval compute the cumulative survival and adjust the # at risk
+      // for the start of the next interval.
+      if (t > currentInterval.getEnd()) {
+        atRisk -= currentInterval.getCensured();
+        float survivors = atRisk - currentInterval.getDied();
+        float tmp = survivors / atRisk;
+        cumulativeSurvival *= tmp;
+
+        // Skip to the next interval
+        atRisk -= currentInterval.getDied();
+        while (intervalIter.hasNext() && t > currentInterval.getEnd()) {
+          currentInterval = intervalIter.next();
+          currentInterval.setCumulativeSurvival(cumulativeSurvival);
+        }
+      }
+
+      val donor = new DonorValue(
+          donors[i].getId(),
+          (String) donors[i].field("donor_vital_status").getValue(),
+          time[i]);
+      currentInterval.addDonor(donor);
+
+      if (!censured[i]) {
+        currentInterval.incDied();
+      }
+    }
+    currentInterval.setCumulativeSurvival(cumulativeSurvival);
+
+    return intervals;
+  }
+
+  private static boolean hasData(SearchHit donor) {
+    val status = (String) donor.field("donor_vital_status").getValue();
+    return status.equalsIgnoreCase("alive") || status.equalsIgnoreCase("deceased");
+  }
+
+  private static Map<UUID, UnionUnit> getEntitySetMap(List<UUID> sets) {
+    val entitySetMapBuilder = new ImmutableMap.Builder<UUID, UnionUnit>();
+
+    for (val id : sets) {
+      val exclusions = sets.stream()
```

Comment:
Heh, maybe this would just work:

```java
return Maps.uniqueIndex(setIds, setId -> new UnionUnit(ImmutableSet.of(setId), Sets.filter(setIds, s -> !s.equal(setId)))));
```